bin/propolis-server/src/lib/vm/request_queue.rs

```diff
@@ -0,0 +1,358 @@
+};
+
+/// An external request made of a VM controller via the server API. Handled by
+/// the controller's state driver thread.
+#[derive(Debug)]
+pub enum ExternalRequest {
+    /// Initializes the VM through live migration by running a
+    /// migration-destination task.
+    MigrateAsTarget {
+        /// The ID of the live migration to use when initializing.
+        migration_id: Uuid,
+
+        /// A handle to the task that will execute the migration procedure.
+        task: tokio::task::JoinHandle<Result<(), MigrateError>>,
+
+        /// The sender side of a one-shot channel that, when signaled, tells the
+        /// migration task to start its work.
+        start_tx: tokio::sync::oneshot::Sender<()>,
+
+        /// A channel that receives commands from the migration task.
+        command_rx: tokio::sync::mpsc::Receiver<MigrateTargetCommand>,
+    },
+
+    /// Resets all the VM's entities and CPUs, then starts the VM.
+    Start,
+
+    /// Asks the state worker to start a migration-source task.
+    MigrateAsSource {
+        /// The ID of the live migration for which this VM will be the source.
+        migration_id: Uuid,
+
+        /// A handle to the task that will execute the migration procedure.
+        task: tokio::task::JoinHandle<Result<(), MigrateError>>,
+
+        /// The sender side of a one-shot channel that, when signaled, tells the
+        /// migration task to start its work.
+        start_tx: tokio::sync::oneshot::Sender<()>,
+
+        /// A channel that receives commands from the migration task.
+        command_rx: tokio::sync::mpsc::Receiver<MigrateSourceCommand>,
+
+        /// A channel used to send responses to migration commands.
+        response_tx: tokio::sync::mpsc::Sender<MigrateSourceResponse>,
+    },
+
+    /// Resets the guest by pausing all devices, resetting them to their
+    /// cold-boot states, and resuming the devices. Note that this is not a
+    /// graceful reboot and does not coordinate with guest software.
+    Reboot,
+
+    /// Halts the VM. Note that this is not a graceful shutdown and does not
+    /// coordinate with guest software.
+    Stop,
+}
+
+/// A set of reasons why a request to queue an external state transition can
+/// fail.
+#[derive(Copy, Clone, Debug, Error)]
+pub enum RequestDeniedReason {
+    #[error("The requested operation requires an active instance")]
+    InstanceNotActive,
+
+    #[error("A migration into this instance is in progress")]
+    MigrationTargetInProgress,
+
+    #[error("The instance is currently starting")]
+    StartInProgress,
+
+    #[error("The instance is already a migration source")]
+    AlreadyMigrationSource,
+
+    #[error(
+        "The requested operation cannot be performed on a migration source"
+    )]
+    InvalidRequestForMigrationSource,
+
+    #[error("The instance is preparing to stop")]
+    HaltPending,
+
+    #[error("The instance failed to start or halted due to a failure")]
+    InstanceFailed,
```

Comment:
a bit nitty, but I think I would prefer if these errors are a bit less verbose, such as removing The/A at the beginning